ListUtil.xs

```diff
@@ -1465,63 +1498,81 @@ PPCODE:
     else if (PL_op->op_private & HINT_STRICT_REFS)
         croak("Can't use string (\"%.32s\") as %s ref while \"strict refs\" in use",
               SvPV_nolen(sub), "a subroutine");
-    else if ((gv = gv_fetchpv(SvPV_nolen(sub), FALSE, SVt_PVCV)))
+    else if ((gv = gv_fetchsv(sub, FALSE, SVt_PVCV)))
         cv = GvCVu(gv);
     if (!cv)
         croak("Undefined subroutine %s", SvPV_nolen(sub));
     if (SvTYPE(cv) != SVt_PVCV && SvTYPE(cv) != SVt_PVFM)
         croak("Not a subroutine reference");
-    for (s = name; *s++; ) {
-        if (*s == ':' && s[-1] == ':')
-            end = ++s;
-        else if (*s && s[-1] == '\'')
-            end = s;
+    for (s = nameptr; s <= nameptr + namelen; s++) {
+        if (s > nameptr && *s == ':' && s[-1] == ':') {
+            end = s - 1;
+            begin = ++s;
+            if (seen_quote)
+                need_subst++;
+        }
+        else if (s > nameptr && *s != '\0' && s[-1] == '\'') {
+            end = s - 1;
+            begin = s;
+            if (seen_quote++)
```

Comment:
Are you quite sure you mean `seen_quote++` and not `++seen_quote` here? I.e. the very first time you see a quote mark you increment `seen_quote` but don't yet set `need_subst` to true, until the second time you see one?

Reply:
If only a single quote character is seen, and no `::` package separator, then the quote is separating the package name from the sub name.  In that case, we don't need to construct a new package name using double colons (which is what needs_substr tracks) because the package name won't have any single quotes (or `::`) in it.  This also explains why the seen_quote check exists in the branch for `::`.

Reply:
Ah, I see. In that case, maybe rather than two variables could we just count the number of quotes seen (named `quotes_seen` so a plural noun suggesting a count), and change the condition to something like

    if(quotes_seen > 1) {
        /* multiple quotes were seen, but only the final one separates the basename from the package.
           the rest need converting into :: */

Reply:
It matters where the quote is seen. `Foo::Bar'baz` is fine, but `Foo'Bar::baz` isn't.

Parsing in C sucks.